File: backend/app/services/hotspot_detector.py
import pandas as pd
import numpy as np
from sklearn.cluster import DBSCAN
from typing import List, Dict, Any
from app.core.config import settings

def detect_hotspots_for_readings(readings: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    Detects spatial pollution hotspots using DBSCAN.
    
    Disclaimer: This identifies spatial clusters of elevated AQI, not necessarily 
    the exact physical source of the pollution.
    """
    if not readings:
        return []
        
    df = pd.DataFrame(readings)
    
    # Filter for valid AQI above threshold
    df_high = df[(df['aqi'].notnull()) & (df['aqi'] >= settings.HOTSPOT_MIN_AQI)].copy()
    
    if len(df_high) < settings.HOTSPOT_MIN_SAMPLES:
        return []

    # Prepare coordinates in radians for Haversine distance
    coords = np.radians(df_high[['latitude', 'longitude']].values)
    
    # eps is radius in radians (earth radius ~ 6371000 meters)
    eps_rad = settings.HOTSPOT_RADIUS_METERS / 6371000.0
    
    db = DBSCAN(eps=eps_rad, min_samples=settings.HOTSPOT_MIN_SAMPLES, algorithm='ball_tree', metric='haversine')
    df_high.loc[:, 'cluster'] = db.fit_predict(coords)
    
    # Exclude noise points (-1)
    clusters = df_high[df_high['cluster'] != -1]
    
    hotspots = []
    
    for cluster_id, group in clusters.groupby('cluster'):
        center_lat = group['latitude'].mean()
        center_lng = group['longitude'].mean()
        
        avg_aqi = group['aqi'].mean()
        peak_aqi = int(group['aqi'].max())
        
        avg_pm25 = group['pm25'].mean()
        peak_pm25 = group['pm25'].max()
        
        avg_pm10 = group['pm10'].mean()
        peak_pm10 = group['pm10'].max()
        
        count = len(group)
        
        from app.services.aqi import get_category_from_aqi
        severity = get_category_from_aqi(peak_aqi)
        
        # Altitude Intelligence
        min_altitude = None
        max_altitude = None
        avg_altitude = None
        if 'altitude' in group.columns:
            valid_altitudes = group['altitude'].dropna()
            if not valid_altitudes.empty:
                min_altitude = float(valid_altitudes.min())
                max_altitude = float(valid_altitudes.max())
                avg_altitude = float(valid_altitudes.mean())
        
        hotspots.append({
            "latitude": float(center_lat),
            "longitude": float(center_lng),
            "radius_meters": settings.HOTSPOT_RADIUS_METERS,
            "average_aqi": float(avg_aqi),
            "peak_aqi": peak_aqi,
            "average_pm25": float(avg_pm25),
            "peak_pm25": float(peak_pm25),
            "average_pm10": float(avg_pm10),
            "peak_pm10": float(peak_pm10),
            "severity": severity,
            "reading_count": count,
            "min_altitude": min_altitude,
            "max_altitude": max_altitude,
            "average_altitude": avg_altitude
        })
        
    return hotspots

File: backend/services/hotspotDetector.js
const { getCategoryFromAQI } = require('./aqi');

const HOTSPOT_MIN_SAMPLES = Number(process.env.HOTSPOT_MIN_SAMPLES || 3);
const HOTSPOT_RADIUS_METERS = Number(process.env.HOTSPOT_RADIUS_METERS || 75);

function haversineDistance(lat1, lon1, lat2, lon2) {
    const R = 6371000; // Earth radius in meters
    const phi1 = (lat1 * Math.PI) / 180;
    const phi2 = (lat2 * Math.PI) / 180;
    const deltaPhi = ((lat2 - lat1) * Math.PI) / 180;
    const deltaLambda = ((lon2 - lon1) * Math.PI) / 180;

    const Math_sin = Math.sin;
    const Math_cos = Math.cos;

    const a =
        Math_sin(deltaPhi / 2) * Math_sin(deltaPhi / 2) +
        Math_cos(phi1) *
            Math_cos(phi2) *
            Math_sin(deltaLambda / 2) *
            Math_sin(deltaLambda / 2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    return R * c;
}

function regionQuery(points, pIndex, eps) {
    const neighbors = [];
    const p1 = points[pIndex];
    for (let i = 0; i < points.length; i++) {
        const p2 = points[i];
        if (haversineDistance(p1.latitude, p1.longitude, p2.latitude, p2.longitude) <= eps) {
            neighbors.push(i);
        }
    }
    return neighbors;
}

function expandCluster(points, labels, pIndex, neighbors, clusterId, eps, minPts) {
    labels[pIndex] = clusterId;
    let i = 0;
    while (i < neighbors.length) {
        const pNeighborIndex = neighbors[i];
        if (labels[pNeighborIndex] === -1) {
            labels[pNeighborIndex] = clusterId; // change from noise to border point
        } else if (labels[pNeighborIndex] === 0) {
            labels[pNeighborIndex] = clusterId;
            const newNeighbors = regionQuery(points, pNeighborIndex, eps);
            if (newNeighbors.length >= minPts) {
                // Not the most efficient but functional for small datasets
                for (const newNeighbor of newNeighbors) {
                    if (!neighbors.includes(newNeighbor)) {
                        neighbors.push(newNeighbor);
                    }
                }
            }
        }
        i++;
    }
}

function dbscan(points, eps, minPts) {
    const NOISE = -1;
    const UNCLASSIFIED = 0;
    let clusterId = 0;
    const labels = new Array(points.length).fill(UNCLASSIFIED);

    for (let i = 0; i < points.length; i++) {
        if (labels[i] !== UNCLASSIFIED) continue;

        const neighbors = regionQuery(points, i, eps);
        if (neighbors.length < minPts) {
            labels[i] = NOISE;
        } else {
            clusterId++;
            expandCluster(points, labels, i, neighbors, clusterId, eps, minPts);
        }
    }
    return labels;
}

function getPriority(val, metric) {
    if (metric === 'pm25') {
        if (val > 250) return 'P1';
        if (val > 150) return 'P2';
        if (val > 55) return 'P3';
    } else if (metric === 'pm10') {
        if (val > 424) return 'P1';
        if (val > 354) return 'P2';
        if (val > 254) return 'P3';
    } else { // default AQI
        if (val > 300) return 'P1';
        if (val > 200) return 'P2';
        if (val > 150) return 'P3';
    }
    return 'Normal';
}

function detectHotspotsForReadings(readings, metric = 'aqi') {
    if (!readings || readings.length === 0) return [];

    let minThreshold = 100; // default for AQI
    if (metric === 'pm25') minThreshold = 35.4; // Unhealthy for Sensitive Groups
    else if (metric === 'pm10') minThreshold = 154; // Unhealthy for Sensitive Groups
    else minThreshold = 100; // AQI

    const highReadings = readings.filter((r) => {
        const val = r[metric];
        return val !== null && val !== undefined && val >= minThreshold;
    });

    if (highReadings.length < HOTSPOT_MIN_SAMPLES) return [];

    const labels = dbscan(highReadings, HOTSPOT_RADIUS_METERS, HOTSPOT_MIN_SAMPLES);

    const clusters = {};
    for (let i = 0; i < highReadings.length; i++) {
        const clusterId = labels[i];
        if (clusterId !== -1) {
            if (!clusters[clusterId]) clusters[clusterId] = [];
            clusters[clusterId].push(highReadings[i]);
        }
    }

    const hotspots = [];

    for (const clusterId in clusters) {
        const group = clusters[clusterId];
        let sumLat = 0, sumLng = 0;
        let sumAqi = 0, maxAqi = -Infinity;
        let sumPm25 = 0, maxPm25 = -Infinity;
        let sumPm10 = 0, maxPm10 = -Infinity;
        
        let minAltitude = Infinity;
        let maxAltitude = -Infinity;
        let sumAltitude = 0;
        let altitudeCount = 0;

        for (const r of group) {
            sumLat += r.latitude;
            sumLng += r.longitude;
            sumAqi += r.aqi;
            if (r.aqi > maxAqi) maxAqi = r.aqi;
            
            if (r.pm25 !== null && r.pm25 !== undefined) {
                sumPm25 += r.pm25;
                if (r.pm25 > maxPm25) maxPm25 = r.pm25;
            }
            if (r.pm10 !== null && r.pm10 !== undefined) {
                sumPm10 += r.pm10;
                if (r.pm10 > maxPm10) maxPm10 = r.pm10;
            }

            if (r.altitude !== null && r.altitude !== undefined) {
                if (r.altitude < minAltitude) minAltitude = r.altitude;
                if (r.altitude > maxAltitude) maxAltitude = r.altitude;
                sumAltitude += r.altitude;
                altitudeCount++;
            }
        }

        const count = group.length;

        hotspots.push({
            latitude: sumLat / count,
            longitude: sumLng / count,
            radius_meters: HOTSPOT_RADIUS_METERS,
            average_aqi: sumAqi / count,
            peak_aqi: maxAqi,
            average_pm25: sumPm25 > 0 ? (sumPm25 / count) : null,
            peak_pm25: maxPm25 !== -Infinity ? maxPm25 : null,
            average_pm10: sumPm10 > 0 ? (sumPm10 / count) : null,
            peak_pm10: maxPm10 !== -Infinity ? maxPm10 : null,
            severity: getCategoryFromAQI(maxAqi),
            priority: getPriority(metric === 'pm25' ? maxPm25 : (metric === 'pm10' ? maxPm10 : maxAqi), metric),
            metric_value: metric === 'pm25' ? maxPm25 : (metric === 'pm10' ? maxPm10 : maxAqi),
            reading_count: count,
            min_altitude: altitudeCount > 0 ? minAltitude : null,
            max_altitude: altitudeCount > 0 ? maxAltitude : null,
            average_altitude: altitudeCount > 0 ? (sumAltitude / altitudeCount) : null
        });
    }

    return hotspots;
}

module.exports = {
    detectHotspotsForReadings
};
